lib/src/simple_op_store.rs

```diff
@@ -84,88 +203,146 @@ fn not_found_to_store_error(err: std::io::Error) -> OpStoreError {
+            simple_op_store_model::View::read_from_in_protocol(&mut protocol).unwrap();
+        Ok(view_from_thrift(&thrift_view))
     }
 
     fn write_view(&self, view: &View) -> OpStoreResult<ViewId> {
         let temp_file = NamedTempFile::new_in(&self.path)?;
 
-        let proto = view_to_proto(view);
-        let mut proto_bytes: Vec<u8> = Vec::new();
-        proto.write_to_writer(&mut proto_bytes)?;
+        let thrift_view = view_to_thrift(view);
+        let mut thrift_bytes: Vec<u8> = Vec::new();
+        let mut protocol = TCompactOutputProtocol::new(&mut thrift_bytes);
+        thrift_view.write_to_out_protocol(&mut protocol)?;
 
-        temp_file.as_file().write_all(&proto_bytes)?;
+        temp_file.as_file().write_all(&thrift_bytes)?;
 
-        let id = ViewId::new(Blake2b512::digest(&proto_bytes).to_vec());
+        let id = ViewId::new(Blake2b512::digest(&thrift_bytes).to_vec());
 
         persist_content_addressed_temp_file(temp_file, self.view_path(&id))?;
         Ok(id)
     }
 
     fn read_operation(&self, id: &OperationId) -> OpStoreResult<Operation> {
         let path = self.operation_path(id);
-        let mut file = File::open(path).map_err(not_found_to_store_error)?;
-
-        let proto: crate::protos::op_store::Operation = Message::parse_from_reader(&mut file)?;
-        Ok(operation_from_proto(&proto))
+        let file = File::open(path).map_err(not_found_to_store_error)?;
+        let mut protocol = TCompactInputProtocol::new(&file);
+        let thrift_operation =
+            simple_op_store_model::Operation::read_from_in_protocol(&mut protocol).unwrap();
+        Ok(operation_from_thrift(&thrift_operation))
     }
 
     fn write_operation(&self, operation: &Operation) -> OpStoreResult<OperationId> {
         let temp_file = NamedTempFile::new_in(&self.path)?;
 
-        let proto = operation_to_proto(operation);
-        let mut proto_bytes: Vec<u8> = Vec::new();
-        proto.write_to_writer(&mut proto_bytes)?;
+        let thrift_view = operation_to_thrift(operation);
+        let mut thrift_bytes: Vec<u8> = Vec::new();
+        let mut protocol = TCompactOutputProtocol::new(&mut thrift_bytes);
+        thrift_view.write_to_out_protocol(&mut protocol)?;
 
-        temp_file.as_file().write_all(&proto_bytes)?;
+        temp_file.as_file().write_all(&thrift_bytes)?;
 
-        let id = OperationId::new(Blake2b512::digest(&proto_bytes).to_vec());
+        let id = OperationId::new(Blake2b512::digest(&thrift_bytes).to_vec());
 
         persist_content_addressed_temp_file(temp_file, self.operation_path(&id))?;
         Ok(id)
     }
 }
 
-fn timestamp_to_proto(timestamp: &Timestamp) -> crate::protos::op_store::Timestamp {
-    let mut proto = crate::protos::op_store::Timestamp::new();
-    proto.millis_since_epoch = timestamp.timestamp.0;
-    proto.tz_offset = timestamp.tz_offset;
-    proto
+fn timestamp_to_thrift(timestamp: &Timestamp) -> simple_op_store_model::Timestamp {
```

Comment:
style: Should this, and the family of similar functions and inverses below, be `From` impls? Failing that, how about methods/associated fns on the model types?

Reply:
> I don't want them on the model because the model is independent of storage backend

By "the model types" I meant e.g. `simple_op_store_model::Timestamp`, which I understand to be a Thrift-specific type that could easily have a `from_whatever` constructor and a `to_whatever` method.

> Do you think that will simplify anything or is it just good practice?

I find associated functions and methods which hang off the type they concern easier to find/navigate than free functions, both with tooling support and manually. I don't think it will have any impact on code complexity. 

Using `From` for infallible constructors specifically is cute and can be convenient sometimes, especially as it permits the terse `.into()` in some cases, but that has to be weighed against being more implicit.

Reply:
Just in case it was unclear, it's possible to add `impl` blocks to a type defined in a crate from any module in that crate. We wouldn't need to edit generated code. `From` is good too though!

Reply:
Ah, makes sense. I had not considered that (but I probably would have guessed that it would work if I had thought about it).